File: justforsidekicks.py
from sys import stdin,stdout

class FTree:
    def __init__(self, f):
        self.n = len(f)
        self.ft = [0 for i in range(self.n + 1)]
        for i in range(1, self.n + 1):
            self.ft[i] += f[i - 1]
            j = i + self.lsone(i)
            if j <= self.n:
                self.ft[j] += self.ft[i]

    def lsone(self, s):
        return s & (-s)

    def query(self, i, j):
        if i > 1:
            return self.query(1, j) - self.query(1, i - 1)
        s = 0
        while j > 0:
            s += self.ft[j]
            j -= self.lsone(j)
        return s

    def update(self, i, v):
        while i <= self.n:
            self.ft[i] += v
            i += self.lsone(i)

    def select(self, k):
        p = 1
        while p << 1 <= self.n: p <<= 1
        i = 0
        while p > 0:
            if k > self.ft[i + p]:
                k -= self.ft[i + p]
                i += p
            p >>= 1
        return i + 1
    
N,Q = map(int,stdin.readline().split())
values = list(map(int,stdin.readline().split()))
gems = list(stdin.readline().strip())
arr = [[],[],[],[],[],[]]
#construct frequency array for every gem type in O(N) time
for char in gems:
    for i in range(6):
        arr[i].append(1) if i == int(char) - 1 else arr[i].append(0)
#construct 6 Fenwick trees, one for each gem type
trees = [FTree(arr[i]) for i in range(6)]
for i in range(Q):
    line = stdin.readline().split()
    a,b = int(line[1]) - 1,int(line[2]) - 1
    if line[0] == "1":
        old = int(gems[a]); old -= 1
        trees[old].update(a + 1, -1)
        trees[b].update(a + 1,1)
        gems[a] = b + 1
    elif line[0] == "2":
        values[a] = b + 1
    elif line[0] == "3":
        cf = [trees[i].query(a + 1, b + 1) for i in range(6)]
        stdout.write(f"{sum(values[i]*cf[i] for i in range(6))}\n")
            

File: justforsidekicks.cpp
#include <bits/stdc++.h>

using namespace std;

#define LSOne(S) ((S) & -(S))                    

typedef long long ll;                            
typedef vector<ll> vll;
typedef vector<int> vi;

class FenwickTree {                              // index 0 is not used
private:
  vll ft;                                        
public:
  FenwickTree(int m) {ft.assign(m + 1, 0);}      

  void build(const vll &f) {
    int m = (int)f.size();                     
    ft.assign(m + 1, 0);
    for (int i = 1; i < m + 1; ++i) {               
      ft[i] += f[i - 1];                             
      if (i + LSOne(i) <= m)                       
        ft[i + LSOne(i)] += ft[i];                 
    }
  }

  FenwickTree(const vll &f) {build(f);}        

  FenwickTree(int m, const vi &s) {              
    vll f(m + 1, 0);
    for (int i = 0; i < (int)s.size(); ++i) ++f[s[i]];                                 
    build(f);                                    
  }

  ll rsq(int j) {                                
    ll sum = 0;
    for (; j; j -= LSOne(j))
      sum += ft[j];
    return sum;
  }

  ll rsq(int i, int j) { return rsq(j) - rsq(i-1); } 

  // updates value of the i-th element by v (v can be +ve/inc or -ve/dec)
  void update(int i, ll v) {
    for (; i < (int)ft.size(); i += LSOne(i))
      ft[i] += v;
  }

  int select(ll k) {                            
    int p = 1;
    while (p*2 < (int)ft.size()) p *= 2;
    int i = 0;
    while (p) {
      if (k > ft[i+p]) {
        k -= ft[i+p];
        i += p;
      }
      p /= 2;
    }
    return i+1;
  }
};

typedef vector<FenwickTree> vft;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    int N,Q; cin >> N >> Q;
    vll values(6,0);
    for (int i = 0; i < 6; i++) cin >> values[i];
    string s; cin >> s;
    vi gems;
    for (int i = 0; i < N; i++) gems.push_back(s[i] - '0' - 1);
    // construct frequency array for every gem type in O(N) time
    vector<vll> arr(6,vll(N,0));
    for (int i = 0; i < N; i++) arr[gems[i]][i] = 1;
    /*
    cout << "test" << "\n";
    for (vll v: arr) {
      for (ll p: v) cout << p << " ";
      cout << "\n";
    } */
    // construct 6 Fenwick trees, one for each gem type
    vft trees;
    for (int i = 0; i < 6; i++) trees.push_back(FenwickTree(arr[i]));
    while (Q--) {
        int q,a,b; cin >> q >> a >> b;
        a--; b--;
        if (q == 1) {
            trees[gems[a]].update(a + 1, -1);
            trees[b].update(a + 1, 1);
            gems[a] = b;
        } else if (q == 2) {
            values[a] = b + 1;
        } else {
            vll cf(6,0);
            for (int i = 0; i < 6; i++) cf[i] = trees[i].rsq(a + 1, b + 1);
            ll sum = 0;
            for (int i = 0; i < 6; i++) sum += (values[i] * cf[i]);
            cout << sum << "\n";
        }
    }
    return 0;
}
